File: src/main/java/plus/maa/backend/controller/response/MaaResult.java
package plus.maa.backend.controller.response;

import com.fasterxml.jackson.annotation.JsonInclude;
import plus.maa.backend.common.MaaStatusCode;

/**
 * @author AnselYuki
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record MaaResult<T>(int statusCode, String message, T data) {
    public static <T> MaaResult<T> success(T data) {
        return success(null, data);
    }

    public static <T> MaaResult<T> success(String msg, T data) {
        return new MaaResult<>(200, msg, data);
    }

    public static <T> MaaResult<T> fail(int code, String msg) {
        return fail(code, msg, null);
    }

    public static <T> MaaResult<T> fail(int code, String msg, T data) {
        return new MaaResult<>(code, msg, data);
    }

    public static MaaResult<Void> fail(MaaStatusCode maaActiveError) {
        return new MaaResult<>(maaActiveError.code, maaActiveError.message, null);
    }

    public static <T> MaaResult<T> fail(MaaStatusCode maaActiveError, T data) {
        return new MaaResult<>(maaActiveError.code, maaActiveError.message, data);
    }
}

File: src/main/kotlin/plus/maa/backend/controller/response/MaaResult.kt
package plus.maa.backend.controller.response

import com.fasterxml.jackson.annotation.JsonInclude

/**
 * @author AnselYuki
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
data class MaaResult<out T>(
    val statusCode: Int,
    val message: String?,
    val data: T?,
) {
    companion object {
        fun <T> success(data: T): MaaResult<T> = success(null, data)

        fun success(): MaaResult<Unit> = success(null, Unit)

        fun <T> success(msg: String?, data: T?): MaaResult<T> = MaaResult(200, msg, data)

        fun fail(code: Int, msg: String?): MaaResult<Nothing> = MaaResult(code, msg, null)
    }
}
